pkg/master/thirdparty/tprregister_controller.go

```diff
@@ -0,0 +1,213 @@
+	for i := 0; i < threadiness; i++ {
+		// runWorker will loop until "something bad" happens.  The .Until will then rekick the worker
+		// after one second
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+
+	// wait until we're told to stop
+	<-stopCh
+}
+
+func (c *tprRegisterController) runWorker() {
+	// hot loop until we're told to stop.  processNextWorkItem will automatically wait until there's work
+	// available, so we don't worry about secondary waits
+	for c.processNextWorkItem() {
+	}
+}
+
+// processNextWorkItem deals with one key off the queue.  It returns false when it's time to quit.
+func (c *tprRegisterController) processNextWorkItem() bool {
+	// pull the next work item from queue.  It should be a key we use to lookup something in a cache
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	// you always have to indicate to the queue that you've completed a piece of work
+	defer c.queue.Done(key)
+
+	// do your work on the key.  This method will contains your "do stuff" logic
+	err := c.syncHandler(key.(schema.GroupVersion))
+	if err == nil {
+		// if you had no error, tell the queue to stop tracking history for your key.  This will
+		// reset things like failure counts for per-item rate limiting
+		c.queue.Forget(key)
+		return true
+	}
+
+	// there was a failure so be sure to report it.  This method allows for pluggable error handling
+	// which can be used for things like cluster-monitoring
+	utilruntime.HandleError(fmt.Errorf("%v failed with : %v", key, err))
+	// since we failed, we should requeue the item to work on later.  This method will add a backoff
+	// to avoid hotlooping on particular items (they're probably still not going to work right away)
+	// and overall controller protection (everything I've done is broken, this controller needs to
+	// calm down or it can starve other useful work) cases.
+	c.queue.AddRateLimited(key)
+
+	return true
+}
+
+func (c *tprRegisterController) enqueueTPR(tpr *extensions.ThirdPartyResource) {
+	_, group, err := thirdpartyresourcedata.ExtractApiGroupAndKind(tpr)
+	if err != nil {
+		utilruntime.HandleError(err)
+		return
+	}
+	for _, version := range tpr.Versions {
+		c.queue.Add(schema.GroupVersion{Group: group, Version: version.Name})
+	}
+}
+func (c *tprRegisterController) handleTPR(groupVersion schema.GroupVersion) error {
+	// check all TPRs.  There shouldn't that many, but if we have problems later we can index them
+	tprs, err := c.tprLister.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	hasResources := false
+	for _, tpr := range tprs {
+		_, group, err := thirdpartyresourcedata.ExtractApiGroupAndKind(tpr)
+		if err != nil {
+			return err
+		}
+		for _, version := range tpr.Versions {
+			if version.Name == groupVersion.Version && group == groupVersion.Group {
+				hasResources = true
+				break
+			}
+		}
+	}
+
+	apiServiceName := groupVersion.Version + "." + groupVersion.Group
+
+	if !hasResources {
```

Comment:
What about if the cache is slow to sync? You could have a new TPR enqueued but it's not in the lister yet?